untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+[Kleene–Rosser paradox]: https://en.wikipedia.org/wiki/Kleene%E2%80%93Rosser_paradox
+[paradoxes-contemporary-logic]: https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi
+
+The lambda symbol used in the standard notation, from which lambda calculus derives its name, most likely was chosen arbitrarily.
+However, there's an apocryphal story the notation was derived from $\hat x$, morphing first into $\Lambda x$, and later into $\lambda x$.
+
+In the mid-1960s, Peter Landin has shown that lambda calculus models arbitrarily complex programming languages.
+This insight arguably kickstarted the research on functional programming languages.
+
+## The pure untyped lambda calculus
+
+The untyped lambda calculus is, first and foremost, a consistent expression rewriting system.
+Calculus that only contains functions and variables is called the pure untyped lambda calculus, "pure" meaning it doesn't have anything beyond what's strictly necessary.
+
+We can formally define the pure untyped lambda calculus in terms of _abstractions_ and _applications_.
+
+_Abstraction_, or functional abstraction, is a definition of a parametric expression, that is to say, a function.
+Lambda calculus only defines univariate functions, but that's easily extended to multivariate, as we'll see later.
+
+We introduce new abstractions using the $\lambda$ symbol.
+The abstraction consists of the _head_ and the _body_, separated by the dot ($.$) symbol.
+The _head_ contains the $\lambda$ symbol and the parameter name.
+The _body_ is an arbitrary expression.
+For example, $\lambda x. x$ is an abstraction.
+As usual in abstract mathematics, we can assign a name to an expression.
+Here we'll be using the equals sign ($=$) for that, for example, $id = \lambda x. x.$
+
+_Application_ is the operation of substituting a specific value for the parameter in the abstraction.
+It is the only operation defined in the pure untyped lambda calculus.
+In the standard notation, the application is denoted simply by whitespace.
+For example, $f\; x$ is the application of $f$ to $x$.
+Since all abstractions are univariate by definition, there isn't a special syntax for multivariate application.
+
+You might notice that using whitespace can lead to ambiguity.
+We'll use parentheses where necessary to avoid it.
+Additionally, we will assume that application is left-associative and binds more tightly than abstractions.
+Hence, $f\;a\;b\;c$ is $((f\;a)\;b)\;c$, i.e. $f$ is applied to $a$, then the result is applied to $b$, then to $c$.
+Since application binds more tightly, we don't need parentheses around the body of an abstraction.
+However, we might need parenthesis around the whole abstraction itself.
+We will also eschew parentheses for nested abstractions, e.g. $\lambda x. \lambda y. x\;y$ is $\lambda x. (\lambda y. (x\;y)).$
+
+Variables in the body of an abstraction that are parameters of said abstraction are called _bound variables_.
```

Comment:
This feels like it goes with "abstraction", so should probably be moved up? 

I believe you could make level-3 subheaders for abstraction (where is everything regarding variables) and application, it would look clearer like that.

Reply:
Good point, moved.

Although I'm not sure about the sub-headers. FWIW I've added them, but it feels kind of meh.